### DISCUSSION TITLE
Add "echo:busy: paused for user" as notification option

### DISCUSSION THREAD

**Discussion by Bendermk6** [berrystephenw/OctoText#52] [Ideas]
When the MMU2S needs user attention you get a "echo:busy: paused for user" from the printer.
I can get a popup alert in the web interface with the "OctoPrint-PrinterAlerts" plugin, but an email or text with OctoText would be much better. I tried different notification levels (error, warning, info) on the plugin, but none have worked to get a email with Octotext.

Is it possible to get a Octotext alert for "echo:busy: paused for user"?

Do you need some log files from the MMU? The "echo:busy: paused for user" should been read only once, because the printer repeat the message till the user get rid of the error on the MMU.

**Comment by berrystephenw** [comment]
Hi, I'm not sure if you are running the release candidate software. There are instructions in the pinned discussion on how change to that branch. I've added pause and resume notices that might work for you. There is one bug that might prevent it from sending a message, but I think I understand it and I will be sending a release (rc3) soon.

**Reply by Bendermk6** [reply]
Hey,
i'm already on 0.2.0rc2.
So far i have no problems with it.
Printername is now in subject, which is nice.

Pause by hand, resuming and cancel all works perfect:
![mail_pause_by_hand](https://user-images.githubusercontent.com/81684925/113984958-f11b4f00-984b-11eb-8153-917b879ad8fa.png)
Only the number "993" when a print is cancelled i a bit confusing. I think it is the printing time in seconds?

The printer don't change the state when in "pause for user", i think that's why OctoText is not triggered.
I've made a short terminal log from a mmu print with a loading error and added some //comments.
[mmu_failed_loading_filament.txt](https://github.com/berrystephenw/OctoText/files/6276820/mmu_failed_loading_filament.txt)

Also my Octoprint.log from that print:
[octoprint.log](https://github.com/berrystephenw/OctoText/files/6276822/octoprint.log)

I hope the log files are helpful.

Thank you.

**Reply by berrystephenw** [reply]
very interesting. Now I see a better picture of what is happening - thanks for those logs. I don't know if I can catch the printer pause with the call backs I have in the code now. I'll look at it closer this evening and see what that would take.

Yes the 993 is the number of seconds, I'll make it more obvious that is what is being reported. Thanks!

**Reply by berrystephenw** [reply]
Ok, I've spent some time playing with the code. One of the issues as you noted is that you only want to send the first message, not all of the ones that you can see on in the logs. 
I've got something now that works but it implements a timeout. It sends the first message it sees, then starts a timer for 10 minutes and doesn't send another one until then. I'm worried if I don't set a timer then it might be possible to get into a state where it never sends another notification. I'm still testing but I'll post up another RC soon.
If it makes sense I can make the timer configurable.

**Reply by Bendermk6** [reply]
Have you had a look at the "OctoPrint-PrinterAlerts" plugin? Maybe there is something useful in it to cover the paused messenges?

I think a configurable timer is a good idea, you can set it to 1 or 2 min, because sometimes a filament change mid print is below 1min -> if it is set to 10 min you maybe have the printer waiting 8min till you get the next message :D

I noticed that the printer also makes the "echo:busy: paused for user" when you load filament and he ask if the color is correct. So maybe the paused message should only be active if the printer is in "printing" mode.

Thanks for your work.

**Reply by berrystephenw** [reply]
Yes on looking at printer alerts - I used the call back code directly for testing. I can put a configurable time out in the settings page. It will be a little harder to limit it to just the printer state but I'll see what I can do!

**Reply by berrystephenw** [reply]
the RC4 update should do what you want. I've limited it to reporting a printer pause for user only when it reports that it is actually printing. The settings menu will allow you to change the timeout. If this works for you, please let me know!

**Comment by Bendermk6** [comment]
Hi,

so i tested three MMU2S loading/unloading fails on the RC4 release. I set the timer to 60 seconds.
All works perfect!

#1 Loading fail:
no filament on FINDA(MMU) and no on PINDA(MK3s) -> after few seconds trying, the printer sends 4 times "pause for user" -> got one mail with:
```
subject: Print Paused by system at 2021-04-12 09:30
text: Print Paused on: Prusa MK3S+ MMU2s printer Message sent from: Prusa MK3S+ MMU2s
```

#2 Loading fail:
filament on FINDA and no on PINDA -> printer is trying 4 times to reload the filament -> printer go into "pause for user" loop, endless echo till user fix the issue -> got one mail every 60sec till i fix the issue.

#3 Unloading fail:
filament on FINDA and no on PINDA -> printer is trying to unload 2 or 3 times -> printer sends 4 times "pause for user" -> got one mail.

After testing and the more information i have now on the load/unload fails, i think setting the timer to 2 - 3 minutes is enough.
I mean, when i have a issue with filament load/unload i'm still there few minutes after fixing it, minimum till the next loading or unloading of the filament.

And i don't now if gmail (or any other service) has some sort "anti-spam-detection" that is triggered when i'm sending one mail every 60  seconds :D

If you need some more terminal/serial logs or testing from the MMU, i can help you out.

Thank you for the work.

**Reply by berrystephenw** [reply]
Excellent! I'm glad it works for you. 

As for Gmail and spam detection, there definitely is some algorithm they use and at some point I expect to see big reports from people saying they stopped seeing emails. Right now that isn't happening!

Thanks for testing this, I'll be folding these changes into the main branch this week.
